lib/tracker.js

```diff
@@ -1,35 +1,172 @@
+const Hyperbee = require('hyperbee')
+const Hyperdrive = require('hyperdrive')
+const Seeders = require('@hyperswarm/seeders')
 const speedometer = require('speedometer')
+const HypercoreId = require('hypercore-id-encoding')
+const mutexify = require('mutexify/promise')
 
 module.exports = class Tracker {
-  constructor () {
-    this.resources = []
+  constructor (store, swarm, opts = {}) {
+    this.store = store
+    this.swarm = swarm
+
+    this.backup = opts.backup === true
+
+    this.lock = mutexify()
+    this.resources = new Map()
+  }
+
+  async add (key, type, opts) {
+    const release = await this.lock()
+    try {
+      await this._add(key, type, opts)
+    } finally {
+      release()
+    }
   }
 
-  add (type, resource, opts) {
+  async _add (key, type, opts) {
+    if (this.has(key)) throw new Error('Resource already added')
+
+    const publicKey = HypercoreId.decode(key)
+    const id = HypercoreId.encode(publicKey)
+
+    // Maybe all this belongs to a "Resource" class, including the internals _createCore/Bee/etc
     const info = {
+      key: id,
       type,
-      [type]: resource,
-      resource,
-      watcher: opts.watcher || null,
-      blocks: { download: speedometer(), upload: speedometer() },
-      network: { download: speedometer(), upload: speedometer() }
+      instance: null,
+      seeders: null,
+      blocks: { down: speedometer(), up: speedometer() },
+      network: { down: speedometer(), up: speedometer() },
+      source: opts.source || null
     }
 
-    if (opts && opts.speedometer) Tracker.speedometer(opts.speedometer, info)
+    if (type === 'core') info.instance = await this._createCore(publicKey, info)
+    else if (type === 'bee') info.instance = await this._createBee(publicKey, info)
+    else if (type === 'drive') info.instance = await this._createDrive(publicKey, info, opts)
+    else if (type === 'seeder') {
+      info.type = 'drive'
+      info.instance = await this._createDrive(publicKey, info, { seeders: true, announce: false })
+    } else if (type === 'list') {
+      for (const r of this.resources) {
+        if (r.type === type) throw new Error('Max lists limit is one')
+      }
+      info.instance = await this._createBee(publicKey, info, { keyEncoding: 'utf-8', valueEncoding: 'json' })
+    }
 
-    this.resources.push(info)
+    this.resources.set(id, info)
 
     return info
   }
 
+  async remove (key) {
+    const info = this.get(key)
+
+    if (info.seeders) await info.seeders.destroy()
+    await info.instance.close()
+
+    this.resources.delete(info.key)
```

Comment:
resources might have been overwritten by the above awaits. check if the resource is what you expect it to be first, `if (this.resources.get(info.key) === info) this.resources.delete(info.key)`

Reply:
I added the lock on remove, I think it's worth it as it keeps the code without extra checks